File: scripts/ci-harness-release.py
#!/usr/bin/env python3
import argparse,hashlib,json,os,shutil,subprocess,tempfile
from pathlib import Path
ROOT=Path(__file__).resolve().parents[1]
def run(cmd,**kw):subprocess.run(cmd,cwd=ROOT,check=True,**kw)
def sha(p):return hashlib.sha256(p.read_bytes()).hexdigest()
def main():
 ap=argparse.ArgumentParser();ap.add_argument('--version',required=True);ap.add_argument('--output',required=True);ap.add_argument('--goose-cli');ap.add_argument('--dry-run-publish',action='store_true');a=ap.parse_args();out=Path(a.output).resolve();shutil.rmtree(out,ignore_errors=True);out.mkdir(parents=True)
 if subprocess.check_output(['git','status','--porcelain'],cwd=ROOT,text=True).strip() and os.environ.get('CI')!='true':raise RuntimeError('release build requires clean git tree')
 run(['python3','scripts/validate-harness-manifests.py']);run(['python3','-m','unittest','discover','-s','scripts/tests/integration','-p','test_harness*.py'])
 with tempfile.TemporaryDirectory(prefix='harness-ci-') as td:
  t=Path(td);ext=t/'external';internal=t/'internal';run(['python3','scripts/resolve-external-skills.py','--staging',str(ext)]);run(['python3','scripts/build-harness.py','--output',str(internal)]);release=out/'release';run(['python3','scripts/assemble-harness-release.py','--internal',str(internal),'--external',str(ext),'--output',str(release),'--version',a.version]);r2=t/'release2';run(['python3','scripts/assemble-harness-release.py','--internal',str(internal),'--external',str(ext),'--output',str(r2),'--version',a.version]);archive=next(release.glob('*.tar'));other=next(r2.glob('*.tar'))
  if archive.read_bytes()!=other.read_bytes():raise RuntimeError('reproducibility failure')
  prefix=t/'prefix';run(['python3','scripts/install-harness-release.py','install','--bundle',str(release),'--prefix',str(prefix)]);run(['python3','scripts/install-harness-release.py','verify','--prefix',str(prefix)])
  result={'schema':'harness-ci-release-v1','version':a.version,'archive':archive.name,'digest':sha(archive),'releaseManifestSha256':sha(release/'release.json'),'lockSha256':sha(ROOT/'harness/external-skills.lock.json'),'dryRunPublish':a.dry_run_publish};(out/'ci-result.json').write_text(json.dumps(result,indent=2,sort_keys=True)+'\n');print(json.dumps(result,sort_keys=True))
if __name__=='__main__':main()

File: src/app/harness-release/src/ci-harness-release.ts
#!/usr/bin/env node
import { mkdtemp, mkdir, readdir, readFile, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { basename, resolve } from "node:path";
import { REPOSITORY_ROOT } from "./repository-root.js";
import { compactJson, isMain, prettyJson, run, sha256File } from "./release-common.js";
import { verifyLocalAttestationFiles } from "./verify-local-attestation.js";
import type { LocalEvaluationBindings } from "@harness/eval-hub/local-evaluation";

export interface CiHarnessReleaseOptions {
  version: string;
  output: string;
  gooseCli?: string | undefined;
  attestation?: string | undefined;
  bindings?: string | undefined;
  profile?: string | undefined;
  dryRunPublish: boolean;
  skipTests: boolean;
  root?: string | undefined;
  ci?: boolean | undefined;
  command?: ((program: string, args: string[]) => Promise<string>) | undefined;
}
export interface CiReleaseResult {
  schema: "harness-ci-release-v1";
  version: string;
  archive: string;
  digest: string;
  releaseManifestSha256: string;
  lockSha256: string;
  dryRunPublish: boolean;
}

function validVersion(version: string) {
  return version !== "" && version !== "." && version !== ".." && !/[\\/]/.test(version);
}

export async function ciHarnessRelease(options: CiHarnessReleaseOptions): Promise<CiReleaseResult> {
  const root = options.root ?? REPOSITORY_ROOT;
  const output = resolve(options.output);
  const command = options.command ?? ((program: string, args: string[]) => run(program, args, { cwd: root, stdout: "inherit" }));
  if (!validVersion(options.version)) throw new Error("invalid release version");
  const inCi = options.ci ?? process.env.CI === "true";
  const status = await run("git", ["status", "--porcelain"], { cwd: root }).catch(() => inCi ? "" : Promise.reject(new Error("release build requires a git worktree outside CI")));
  if (status.trim() !== "" && !inCi) throw new Error("release build requires clean git tree");

  await rm(output, { recursive: true, force: true });
  await mkdir(output, { recursive: true });
  await command("pnpm", ["--dir", "src/app", "--filter", "@harness/harness-release", "build"]);
  await command("node", ["src/app/harness-release/dist/validate-harness-manifests.js"]);
  if (!options.skipTests) {
    await command("pnpm", ["--dir", "src/app", "--filter", "@harness/harness-release", "test"]);
  }

  const temporary = await mkdtemp(resolve(tmpdir(), "harness-ci-"));
  try {
    const external = resolve(temporary, "external");
    const internal = resolve(temporary, "internal");
    const release = resolve(output, "release");
    const secondRelease = resolve(temporary, "release2");
    await command("node", ["src/app/harness-release/dist/resolve-external-skills.js", "--staging", external]);
    await command("node", ["src/app/harness-release/dist/build-harness.js", "--output", internal]);
    for (const destination of [release, secondRelease]) {
      await command("node", ["src/app/harness-release/dist/assemble-harness-release.js", "--internal", internal, "--external", external, "--output", destination, "--version", options.version]);
    }
    const archiveName = (await readdir(release)).find((name) => name.endsWith(".tar"));
    const secondArchiveName = (await readdir(secondRelease)).find((name) => name.endsWith(".tar"));
    if (archiveName === undefined || secondArchiveName === undefined) throw new Error("release archive missing");
    const archive = resolve(release, archiveName);
    const secondArchive = resolve(secondRelease, secondArchiveName);
    if (!(await readFile(archive)).equals(await readFile(secondArchive))) throw new Error("reproducibility failure");

    const digest = await sha256File(archive);
    if (!options.dryRunPublish) {
      if (options.attestation === undefined || options.bindings === undefined || options.profile === undefined) throw new Error("publication requires --attestation, --bindings, and --profile");
      const suppliedBindings = JSON.parse(await readFile(resolve(options.bindings), "utf8")) as LocalEvaluationBindings;
      if (suppliedBindings.release !== digest) throw new Error("attestation bindings release digest does not match exact candidate");
      await verifyLocalAttestationFiles({ attestation: resolve(options.attestation), bindings: resolve(options.bindings), profile: resolve(options.profile) });
    }
    const result: CiReleaseResult = {
      schema: "harness-ci-release-v1",
      version: options.version,
      archive: basename(archive),
      digest,
      releaseManifestSha256: await sha256File(resolve(release, "release.json")),
      lockSha256: await sha256File(resolve(root, "src/harness/external-skills.lock.json")),
      dryRunPublish: options.dryRunPublish,
    };
    await writeFile(resolve(output, "ci-result.json"), prettyJson(result));
    return result;
  } finally {
    await rm(temporary, { recursive: true, force: true });
  }
}

function usage() {
  return "usage: ci-harness-release --version VERSION --output OUTPUT [--goose-cli GOOSE_CLI] [--attestation FILE --bindings FILE --profile FILE] [--dry-run-publish] [--skip-tests]";
}
export function parseCiArgs(argv: string[]): CiHarnessReleaseOptions | "help" {
  let version: string | undefined;
  let output: string | undefined;
  let gooseCli: string | undefined;
  let attestation: string | undefined;
  let bindings: string | undefined;
  let profile: string | undefined;
  let dryRunPublish = false;
  let skipTests = false;
  for (let index = 0; index < argv.length; index++) {
    const argument = argv[index]!;
    if (argument === "-h" || argument === "--help") return "help";
    if (argument === "--dry-run-publish") { dryRunPublish = true; continue; }
    if (argument === "--skip-tests") { skipTests = true; continue; }
    if (["--version", "--output", "--goose-cli", "--attestation", "--bindings", "--profile"].includes(argument)) {
      const value = argv[++index];
      if (value === undefined || value.startsWith("--")) throw new Error("argument " + argument + ": expected one argument");
      if (argument === "--version") version = value;
      else if (argument === "--output") output = value;
      else if (argument === "--goose-cli") gooseCli = value;
      else if (argument === "--attestation") attestation = value;
      else if (argument === "--bindings") bindings = value;
      else profile = value;
      continue;
    }
    throw new Error("unrecognized arguments: " + argument);
  }
  if (version === undefined || output === undefined) throw new Error("the following arguments are required:" + (version === undefined ? " --version" : "") + (output === undefined ? " --output" : ""));
  return { version, output, gooseCli, attestation, bindings, profile, dryRunPublish, skipTests };
}
export async function main(argv = process.argv.slice(2)) {
  const options = parseCiArgs(argv);
  if (options === "help") { console.log(usage()); return; }
  console.log(compactJson(await ciHarnessRelease(options)).trimEnd());
}
if (isMain(import.meta.url)) main().catch((error: unknown) => { console.error(error instanceof Error ? error.message : error); process.exitCode = 1; });